File: leetcode/solutions/1008.ConstructBinarySearchTreeFromPreorderTraversal/bstFromPreorder.go
package solution

import (
	"github.com/yukisato/leetcode-util/datastructure"
)

// TreeNode structure for LeetCode
type TreeNode = datastructure.TreeNode

func bstFromPreorder(preorder []int) *TreeNode {
	root := &TreeNode{}
	buildBst(root, preorder)

	return root
}

func buildBst(root *TreeNode, nums []int) {
	root.Val = nums[0]
	end := 1

	for ; end < len(nums) && nums[0] > nums[end]; end++ {
	}

	if end > 1 {
		root.Left = &TreeNode{}
		buildBst(root.Left, nums[1:end])
	}
	if len(nums) > end {
		root.Right = &TreeNode{}
		buildBst(root.Right, nums[end:])
	}
}

File: leetcode/solutions/1008.ConstructBinarySearchTreeFromPreorderTraversal/bstFromPreorder.ts
import { BinaryTreeNode } from "../../../datastructure/BinaryTreeNode.ts";

export default function bstFromPreorder(
  preorder: number[]
): BinaryTreeNode<number> | null {
  if (preorder.length === 0) {
    return null;
  }

  const root: BinaryTreeNode<number> = {
    val: preorder[0],
    left: null,
    right: null,
  };
  let i = 1;

  while (root.val > preorder[i] && i < preorder.length) {
    i++;
  }

  root.left = bstFromPreorder(preorder.slice(1, i));
  root.right = bstFromPreorder(preorder.slice(i));

  return root;
}
